clients/application-kafka/src/main/java/org/eclipse/hono/application/client/kafka/impl/KafkaBasedCommandSender.java

```diff
@@ -157,4 +265,132 @@ public KafkaBasedCommandSender(
+                    })
+                    .map(consumer -> {
+                        LOGGER.debug("created command response consumer or tenant[{}]", tenantId);
+                        span.log("created command response consumer");
+                        commandResponseSubscriptions.put(tenantId, consumer);
+                        return null;
+                    });
+        }
+    }
+
+    /**
+     * Wrapped promise with an expiration mechanism, failing the promise after a given time if it has not been completed
+     * yet.
+     *
+     * @param <T> The type of the result that the code produces.
+     */
+    private class ExpiringCommandPromise<T> {
+        private final Promise<T> promise = Promise.promise();
+        private final Span span;
+        private Long timerId;
+
+        /**
+         * Starts a timer so that after the given timeout value, this promise shall get failed if not completed already.
+         *
+         * @param timeout The timeout duration to use for the timer.
+         * @param timeOutHandler The operation to run after this promise got failed as part of a timeout.
+         * @param span The active OpenTracing span for this operation. It is not to be closed in this method!
+         *             An implementation should log (error) events on this span and it may set tags and use this span as
+         *             the parent for any spans created in this method.
+         * @throws NullPointerException if the timeout or span is {@code null}
+         */
+        ExpiringCommandPromise(final Duration timeout, final Handler<Void> timeOutHandler, final Span span) {
+            Objects.requireNonNull(timeout);
+            Objects.requireNonNull(span);
+
+            this.span = span;
+            timerId = vertx.setTimer(timeout.toMillis(), id -> {
+                final SendMessageTimeoutException error = new SendMessageTimeoutException(
+                        "send command/wait for response timed out after " + timeout + "ms");
+                timerId = null;
+                LOGGER.debug("cancelling sending command and waiting for response after {} ms", timeout);
+                TracingHelper.logError(span, error);
+                promise.tryFail(error);
+                Optional.ofNullable(timeOutHandler)
+                        .ifPresent(handler -> handler.handle(null));
+            });
+        }
+
+        /**
+         * Completes this promise with the given result and stops the expiration timer.
+         *
+         * @param result The result to complete this promise with.
+         * @throws NullPointerException if the result is {@code null}
+         */
+        final void tryCompleteAndCancelTimer(final AsyncResult<T> result) {
+            Objects.requireNonNull(result);
+
+            Optional.ofNullable(timerId)
+                    .ifPresent(vertx::cancelTimer);
+
+            if (result.succeeded()) {
+                LOGGER.trace("received command response");
```

Comment:
It would be good to log the correlation id here.

Reply:
I have pushed the changes related to the above comments.